### ISSUE TITLE
Unable to load DLLs with transitive dependencies

### ISSUE THREAD

**Issue by bclothier** [WaynePhillipsEA/twinbasic#455]
**Describe the bug**
It seems that when using DLLs that has their own dependencies, the dependent DLL do not get loaded and thus tB fails with a runtime error regarding not finding a DLL entry point. Given that other executables works with the same DLL, I'm thinking that twinBasic should be able to load the dependent DLLs in the same folder where it loaded the entry DLL from.

**To Reproduce**
1. [Download the ICU library](https://github.com/unicode-org/icu/releases/tag/release-69-1) and install. In this example, I used `C:\icu\bin` to store the DLLs from the package.
2. In a twinBasic project, add the following module:
```
    Private Declare PtrSafe Function u_strlen CDecl Lib "C:\icu\bin\icuuc69.dll" Alias "u_strlen_69" (ByRef s As Integer) As Long
    Public Sub Test_Call_Dll_With_Dependencies()
        Dim s As Integer
        s = 1
        Debug.Print u_strlen(s)
    End Sub
```
3. Run it. You will get a hidden messagebox (it pops behind VSC):
![image](https://user-images.githubusercontent.com/2367644/135382297-ca9688e4-faab-4f10-b3ff-af5e28e19055.png)
4. Close VSC.
5. Go to system/advanced system properties/environmental variables then edit the PATH environmental variable and add to the end: `;C:\icu\bin`
6. Restart VSC and open the same project.
7. Run the same procedure. You will get: `1` printed via the debug console.

**Expected behavior**
The program should be able to run without having to add the DLL's path to the PATH environmental variable. The transitive dependents are all in the same folder so we already know here to find it via the fully-qualified `Lib` clause. I assume that those transitive DLLs are not searched using the first DLL's path.

**IMPORTANT**
[According to Microsoft, ICU library are available by default on Windows 10. Therefore, this bug report may not reproduce on Windows 10 that already has the ICU binaries shipped.](https://docs.microsoft.com/en-us/windows/win32/intl/international-components-for-unicode--icu-)

**Desktop (please complete the following information):**
 - OS: Windows 7
 - twinBASIC compiler version v0.10.5788

**Comment by WaynePhillipsEA** [comment]
Thanks for reporting, @bclothier.   I haven't looked at this yet, but this is likely simply due to tB not setting the current folder (using  SetCurrentDirectory) before calling LoadLibrary, so should be an easy fix when I get to it.

**Comment by Kr00l** [comment]
I guess SetCurrentDirectory is not applicable as it affects the whole process.
Maybe the issue is resolved automatically once tB  late bounds declare via LoadLibrary.

**Comment by bclothier** [comment]
FWIW, the [`LoadLibraryEx`](https://docs.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-loadlibraryexa) has a flag `LOAD_LIBRARY_SEARCH_DLL_LOAD_DIR ` that seems relevant here. Furthermore, the remarks mentions [`AddDllDirectory`](https://docs.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-adddlldirectory) for customizing the search paths.

**Comment by Kr00l** [comment]
No flags on LoadLibraryEx is equal to LoadLibrary.
I would keep it as LoadLibrary unless a new **Attribute** is used on the `Declare` where a number can be passed as flags. If the flag attribute exists use LoadLibraryEx else LoadLibrary.

**Comment by bclothier** [comment]
I think we're talking about 2 different things. You seem to be referring to us creating a `Declare` statement to invoke the `LoadLibrary*` API where I was suggesting that tB's internal implementation can use `LoadLibraryEx` to avoid a call to `SetCurrentDirectory`.

**Comment by Kr00l** [comment]
So you suggest using LoadLibraryEx always instead of LoadLibrary?
Or how should the declare know which to use without an additional attribute?

**Comment by bclothier** [comment]
That'd be up to Wayne to decide how to implement the handling of runtime library loading behind the `Declare` statement. Because the `Lib` can be a fully-qualified path, it needs to be able to handle the case where the path is not relative and thus might not be in one of the search paths normally used when loading the library.

**Comment by Kr00l** [comment]
There is also `SetDllDirectory` which is supported on older OS where `AddDllDirectory` is only as of Windows 8.
So (just for a suggestion) tB could call SetDllDirectory, LoadLibrary and then SetDllDirectory with NULL to restore default.

The SetDllDirectory may be needed only for special cases like this. So how to avoid calling it always for all API declares and only where necessary? Via an attribute?

**Comment by bclothier** [comment]
There are API that can tell you whether a string is a relative path or an absolute path. Parsing the `Lib` should be sufficient since you can infer that if it's an absolute path or perhaps even the case where there's a folder (e.g. `... Lib "foo\bar.dll" ...`), then handle additional flags/calls to ensure that the targeted dll's directory is included in the search path.

**Comment by Kr00l** [comment]
I am wondering how VB6 behaves. It could be tested with the [VBCDeclFix](https://www.vbforums.com/showthread.php?890388-VB6-VBCDeclFix-The-Add-in-allows-you-to-use-Cdecl-functions-in-VB6-IDE) add-in...

**Comment by bclothier** [comment]
I’m away but you can certainly test using the library and code in the OP.

Alternatively, you could test using [those zip dlls](http://www.kallal.ca/zip/index.htm), which would not require cdecl calling convention. I haven’t tested with this so don’t know for a fact there is a transitive dependency between the dll files. That can be verified using an utility like dependency walker.

**Comment by WaynePhillipsEA** [comment]
> There are API that can tell you whether a string is a relative path or an absolute path. Parsing the `Lib` should be sufficient since you can infer that if it's an absolute path or perhaps even the case where there's a folder (e.g. `... Lib "foo\bar.dll" ...`), then handle additional flags/calls to ensure that the targeted dll's directory is included in the search path.

I'm concerned that this would change VBx compatibility behaviour, therefore I will introduce a new attribute for this change.

**Comment by WaynePhillipsEA** [comment]
This feature is now available as of v0.13.41.  Introduced a `[ SetDllDirectory ]` boolean attribute that can be used before declare statements, that when set causes [SetDllDirectory](https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-setdlldirectoryw) to be used before (& after) the LoadLibrary call made by tB to allow for transitive DLL dependencies to be loaded from the same path as the root DLL.
